### DISCUSSION TITLE
posts/upgrade-winget-itself/

### DISCUSSION THREAD

**Discussion by giscus** [at-wr/blog#3] [Blog Posts Comments]
# posts/upgrade-winget-itself/

This tutorial will show you how to upgrade winget itself on Windows without uninstalling the old version.

https://wrye.dev/posts/upgrade-winget-itself/

**Comment by bellemam** [comment]
I had to use 
`Add-AppxPackage -Path '[Path_to_.msixbundle]' -ForceApplicationShutdown`
in order to successfully update the package

**Reply by at-wr** [reply]
Thanks for the advice!

**Reply by Apryed** [reply]
Fastest way is to directly get the lastest version - Which is not preview - and ForceApplicationShutdown like @bellemam said:
`Add-AppxPackage -Path "https://github.com/microsoft/winget-cli/releases/latest/download/Microsoft.DesktopAppInstaller_8wekyb3d8bbwe.msixbundle" -ForceApplicationShutdown`

If they change the package name - which they won't - we should have to check back at github.

**Comment by barto28** [comment]
Very helpful thank you :)

**Comment by vdias** [comment]
# Function to check if winget is installed and get its version
function Get-InstalledWingetVersion {
    try {
        $wingetVersion = winget --version 2>$null
        if ($wingetVersion) {
            Write-Host "Winget is installed. Version: $wingetVersion"
            return $wingetVersion
        } else {
            Write-Host "Winget is not installed on this system." -ForegroundColor Red
            return $null
        }
    } catch {
        Write-Host "Error checking Winget installation." -ForegroundColor Red
        return $null
    }
}

# Function to get the latest release version of Winget from GitHub
function Get-LatestWingetVersion {
    $url = "https://api.github.com/repos/microsoft/winget-cli/releases/latest"
    $releaseInfo = Invoke-RestMethod -Uri $url
    $latestVersion = $releaseInfo.tag_name -replace "v", ""  # Remove the 'v' from the version
    Write-Host "Latest Winget version on GitHub: $latestVersion"
    return $latestVersion
}

# Function to download and install the latest Winget release from GitHub
function Install-LatestWinget {
    $downloadUrl = Get-LatestWingetRelease
    if ($downloadUrl) {
        $wingetInstallerPath = "$env:TEMP\winget-latest.msixbundle"
        Write-Host "Downloading the latest Winget release from GitHub..."
        Invoke-WebRequest -Uri $downloadUrl -OutFile $wingetInstallerPath
        
        Write-Host "Installing Winget..."
        Add-AppxPackage -Path $wingetInstallerPath
        Write-Host "Winget installation completed."
    }
}

# Function to get the download URL for the latest Winget release
function Get-LatestWingetRelease {
    $url = "https://api.github.com/repos/microsoft/winget-cli/releases/latest"
    $releaseInfo = Invoke-RestMethod -Uri $url
    $latestAsset = $releaseInfo.assets | Where-Object { $_.name -like "*.msixbundle" }
    
    if ($latestAsset -ne $null) {
        return $latestAsset.browser_download_url
    } else {
        Write-Host "Error: Could not find the latest Winget release on GitHub." -ForegroundColor Red
        return $null
    }
}

# Main script execution
$installedVersion = Get-InstalledWingetVersion
$latestVersion = Get-LatestWingetVersion

if ($installedVersion -ne $null -and $latestVersion -ne $null) {
    if ($installedVersion -lt $latestVersion) {
        Write-Host "A newer version of Winget is available. Installing version $latestVersion..."
        Install-LatestWinget
    } else {
        Write-Host "You already have the latest version of Winget installed."
    }
} else {
    Write-Host "Installing the latest version of Winget from GitHub..."
    Install-LatestWinget
}

**Reply by at-wr** [reply]
This is amazing, thanks!

**Reply by dg-wfk** [reply]
Wo, thanks a lot.
Here is the complete code for easy copying and I put the main routine in a function called "Update-LatestWinget":
```
function Get-InstalledWingetVersion {
  try {
    $wingetVersion = winget --version 2>$null
    if ($wingetVersion) {
      Write-Host "Winget is installed. Version: $wingetVersion"
      return $wingetVersion
    } else {
      Write-Host "Winget is not installed on this system." -ForegroundColor Red
      return $null
    }
  } catch {
    Write-Host "Error checking Winget installation." -ForegroundColor Red
    return $null
  }
}

function Get-LatestWingetVersion {
  $url = "https://api.github.com/repos/microsoft/winget-cli/releases/latest"
  $releaseInfo = Invoke-RestMethod -Uri $url
  $latestVersion = $releaseInfo.tag_name -replace "v", "" # Remove the 'v' from the version
  Write-Host "Latest Winget version on GitHub: $latestVersion"
  return $latestVersion
}

function Install-LatestWinget {
  $downloadUrl = Get-LatestWingetRelease
  if ($downloadUrl) {
    $wingetInstallerPath = "$env:TEMP\winget-latest.msixbundle"
    Write-Host "Downloading the latest Winget release from GitHub..."
    Invoke-WebRequest -Uri $downloadUrl -OutFile $wingetInstallerPath
    Write-Host "Installing Winget..."
    Add-AppxPackage -Path $wingetInstallerPath
    Write-Host "Winget installation completed."
  }
}

function Get-LatestWingetRelease {
  $url = "https://api.github.com/repos/microsoft/winget-cli/releases/latest"
  $releaseInfo = Invoke-RestMethod -Uri $url
  $latestAsset = $releaseInfo.assets | Where-Object { $_.name -like "*.msixbundle" }
  if ($latestAsset -ne $null) {
    return $latestAsset.browser_download_url
  } else {
    Write-Host "Error: Could not find the latest Winget release on GitHub." -ForegroundColor Red
    return $null
  }
}

function Update-LatestWinget {
  $installedVersion = Get-InstalledWingetVersion
  $latestVersion = Get-LatestWingetVersion
  if ($installedVersion -ne $null -and $latestVersion -ne $null) {
    if ($installedVersion -lt $latestVersion) {
      Write-Host "A newer version of Winget is available. Installing version $latestVersion..."
      Install-LatestWinget
    } else {
      Write-Host "You already have the latest version of Winget installed."
    }
  } else {
    Write-Host "Installing the latest version of Winget from GitHub..."
    Install-LatestWinget
  }
}
```
